platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/BundleMigrator.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.admin.application.service.migratable;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility class capable of managing bundles during export or import. */
+public class BundleMigrator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BundleMigrator.class);
+
+  private final BundleProcessor processor;
+
+  @VisibleForTesting
+  BundleMigrator(BundleProcessor processor) {
+    this.processor = processor;
+  }
+
+  public BundleMigrator() {
+    this(new BundleProcessor());
+  }
+
+  /**
+   * Exports all bundles in memory into their corresponding Json format.
+   *
+   * @return a list of the Json bundles to export
+   */
+  public List<JsonBundle> exportBundles() {
+    return Stream.of(processor.listBundles(getBundleContext()))
+        .map(JsonBundle::new)
+        .collect(Collectors.toList()); // preserve order
+  }
+
+  /**
+   * Import the specified bundles.
+   *
+   * <p>The implementation will loop until it fails to find a bundle in memory or it fails to change
+   * the state of a bundle or again until the state of the bundles in memory matches the state of
+   * the bundles on the original system.
+   *
+   * @param report the report where to record errors
+   * @param jprofile the profile from which to import exported bundles
+   * @return <code>true</code> if all were imported successfully; <code>false</code> otherwise
+   */
+  public boolean importBundles(ProfileMigrationReport report, JsonProfile jprofile) {
+    final BundleContext context = getBundleContext();
+    final TaskList tasks = newTaskList(report);
+
+    // loop until we can determine that all bundles that should be started. stopped, installed, or
+    // uninstalled are or until we get an error
+    while (true) {
```

Comment:
❓ `while (true)` loops always make me nervous, especially in this case where there's no default `else` in the `if/else if` statement inside the loop. Do you think calling `processBundles()` once before to loop to initialize `tasks` and using `while (!tasks.isEmpty())` would be safer?

Reply:
We need to call `processBundles()` within the loop as we want to re-process everything after each iteration cause each iteration will have make changes to memory. I did update the `execute()` documentation to clarify that the task list is cleared evert time after execution.

The TaskList's execution is designed to return false if it exceeds the maximum number of attempts or if it failed to actually execute a task in which case we will bail out of the loop with false. We will also bail out of the loop if the task list ends up being empty after processing. As such, we will be safe in this particular case.